### DISCUSSION TITLE
What tech stack you are using in Pocketbase?

### DISCUSSION THREAD

**Discussion by tazim404** [pocketbase/pocketbase#608] [Q&A]
Which go framework you are using in the backend?
And how did you embed the svelte app with go?

**Comment by ganigeorgiev** [comment]
[Echo](https://github.com/labstack/echo/tree/v5_alpha) is used for routing, if that's what you are asking. You could find all dependencies in the go.mod - https://github.com/pocketbase/pocketbase/blob/master/go.mod.

For the svelte app embedding we are not doing anything special and use the std go `embed` package - https://pkg.go.dev/embed.
You could find the go file here - https://github.com/pocketbase/pocketbase/blob/master/ui/embed.go.

**Reply by tazim404** [reply]
One last question. Does this project use any orm or you are executing operation by writing SQL statments.

**Reply by ganigeorgiev** [reply]
We use a query builder - https://pkg.go.dev/github.com/pocketbase/dbx (fork of ozzo-dbx with better sqlite support).

**Reply by tazim404** [reply]
Sorry to bother you again but I have two doubts:

- How you are handling redirecting to /login and /installer when no admin is available without making any request to the server to check whether any admin is available or not
- How you are handling the migration of database tables?

**Reply by ganigeorgiev** [reply]
1. You could find the redirect in https://github.com/pocketbase/pocketbase/blob/master/apis/base.go#L173-L214. There is also this middleware - https://github.com/pocketbase/pocketbase/blob/master/apis/middlewares.go#L81-L101, that is attached to the admins/create handler.

2. We have a simple migration runner, that could be found in https://github.com/pocketbase/pocketbase/blob/master/tools/migrate/runner.go. Basically each migration is a Go file that is registered to a `MigrationsList`. Practical example how to create migrations when using PocketBase as framework could be found in https://pocketbase.io/docs/db-migrations/.

**Reply by jescoop** [reply]
@ganigeorgiev is this compatible with net/http like gorilla/mux or go-chi/chi?

```go
package main

import (
	"fmt"
	"net/http"
	"time"
)

func greet(w http.ResponseWriter, r *http.Request) {
	fmt.Fprintf(w, "Hello World! %s", time.Now())
}

func main() {
	http.HandleFunc("/", greet)
	http.ListenAndServe(":8080", nil)
}
```

**Reply by ganigeorgiev** [reply]
@jescoop PocketBase router (echo) internally uses `net/http` but the handler func definition is slightly different since echo also has a centralized error handling and middlewares support (_to be technically correct, you can convert it to its final `net/http` form but that's not how it was intended to be used_; you could also check this issue https://github.com/labstack/echo/issues/1781).

**Comment by tazim404** [comment]
I get how you are updating/creating a new column in the table but how you are updating/creating a table with the relationship?

**Reply by ganigeorgiev** [reply]
I'm not sure that I understand what you are asking. If you are looking for how the `relation` type field is stored - we are storing only the id(s) of the related record (for single relations as plain string and for multiple - as json array). When querying over relation columns we are normalizing it with the help of sqlite `json_each` in https://github.com/pocketbase/pocketbase/blob/master/resolvers/record_field_resolver.go#L215-L216.
The foreign-key constraint and the cascade operations are handled on application level (this may change in the future).

Could you elaborate a little more on your use case?

**Reply by tazim404** [reply]
My question is simple how you are creating a column that has a relationship

[https://github.com/pocketbase/pocketbase/blob/838ed661ce398c36d533af840233a6d9b1e5da18/daos/record.go#L278](https://github.com/pocketbase/pocketbase/blob/838ed661ce398c36d533af840233a6d9b1e5da18/daos/record.go#L278.)
Here I am not able to see any relationship-related properties.

And I am running the application in debug mode but it does not log any relationship key constrain query

**Reply by ganigeorgiev** [reply]
I'm sorry. I'm still not able to understand. What "relationship-related properties" are you expecting? Could you provide a code sample what you are trying to do?

**Comment by tazim404** [comment]
Let's say I want to add a new text column in the table then the application simply makes the PUT request and then I updated the collection column after that [SyncRecordTableSchema](https://github.com/pocketbase/pocketbase/blob/838ed661ce398c36d533af840233a6d9b1e5da18/daos/record.go#L278) function run which alters the ADD column in the table. And I can see the query logs in the terminal while running the application in debug mode.

Now let's take the second scenario where I want to add a new column that has a relationship with another column in a different table which again runs [SyncRecordTableSchema](https://github.com/pocketbase/pocketbase/blob/838ed661ce398c36d533af840233a6d9b1e5da18/daos/record.go#L278) function. So here my doubt arises that how you are differentiating columns whether it is a relationship column or a normal column (like text, number, etc), and is there any special function that you defined that creates a relationship column?

At last one personal question. How many years of experience do you have  Gani Georgiev in Coding?

**Reply by ganigeorgiev** [reply]
The column definition of each field is determined by https://github.com/pocketbase/pocketbase/blob/838ed661ce398c36d533af840233a6d9b1e5da18/models/schema/schema_field.go#L87-L99.

I'm still not sure what you are trying to achieve but If you are asking for how to create collections programmatically:

```go
someRule := "@request.user.id != ''"

collection := &models.Collection{
    Name:       "test",
    CreateRule: &someRule,
    ListRule:   &someRule,
    ViewRule:   &someRule,
    UpdateRule: &someRule,
    DeleteRule: &someRule,
    Schema: schema.NewSchema(
        &schema.SchemaField{
            Id:   "field1",
            Name: "title",
            Required: true,
            Type: schema.FieldTypeText,
            Options: &schema.TextOptions{},
        },
        &schema.SchemaField{
            Id:       "field2",
            Name:     "some_relation",
            Type:     schema.FieldTypeRelation,
            Options: &schema.RelationOptions{
                CollectionId: "some_collection_id",
                MaxSelect:     2,
                CascadeDelete: true,
            },
        },
    ),
}

// this will save the above collection and will automatically create/update the related records table
if err := app.Dao().SaveCollection(collection); err != nil {
    return err;
}
```

**Reply by tazim404** [reply]
I am just asking where you are building a query to create or update a column with the relationship.

**Reply by ganigeorgiev** [reply]
OK, create/update a column and create/update a column value are 2 totally different things.

For creating/updating the column, as in the table schema - this is located in the `SyncRecordTableSchema` method.
Create:
https://github.com/pocketbase/pocketbase/blob/838ed661ce398c36d533af840233a6d9b1e5da18/daos/record.go#L289-L292
Update:
https://github.com/pocketbase/pocketbase/blob/838ed661ce398c36d533af840233a6d9b1e5da18/daos/record.go#L351

---

For creating/updating each record field value (with validations) - the RecordUpsert form takes care for that in https://github.com/pocketbase/pocketbase/blob/master/forms/record_upsert.go#L360-L402.

I'm still not able to understand what you are trying to do. Please elaborate on your issue or provide some code example, otherwise we are just writing the same thing over and over.

**Reply by tazim404** [reply]
My last attempt to ask this question.

See I understood how you are updating/creating basic data-type columns (like Text, Bool, Number, etc)  in an existing table. The thing that I want to know while creating a column in an existing table is what type of validation, and different function you are running before committing to DB
![Screenshot from 2022-10-05 01-08-45](https://user-images.githubusercontent.com/68907772/193910562-395deefb-ca70-4c04-8e97-93667e0240ee.png)

**Reply by ganigeorgiev** [reply]
> The thing that I want to know while creating a column in an existing table is what type of validation, and different function you are running before committing to DB.

Validation and different functions for what exactly - the collection definition, the records table or the record itself?

For create/update collections (aka. the meta data of your records table) we have `forms.CollectionUpsert` -https://github.com/pocketbase/pocketbase/blob/master/forms/collection_upsert.go

For keeping the records table schema in sync with the collections - we use the `SyncRecordTableSchema` method - https://github.com/pocketbase/pocketbase/blob/838ed661ce398c36d533af840233a6d9b1e5da18/daos/record.go#L278-L377

For create/update of records we use `forms.RecordUpsert` - https://github.com/pocketbase/pocketbase/blob/master/forms/record_upsert.go. The base data validators for each field types are available in https://github.com/pocketbase/pocketbase/blob/master/forms/validators/record_data.go

**Reply by tazim404** [reply]
After opening the database file in the SQLite browser I understood how I was working and I got my answer from expand query.

One thing more what is the difference between profile and users in DB.

**Reply by ganigeorgiev** [reply]
`profiles` currently is a system collection intended to hold your custom users data (eg. name, avatar, etc.). Each profile is associated to 1 user.

`users` are the accounts of your application. They have fixed fields like email, verified, passwordHash, etc. that usually comes with some special handling depending who is making the request.

In the next v0.8.0 release the `user` and `relation` type fields will be merged per #376.

**Comment by tazim404** [comment]
Can you please elaborate on how `?expand=rel1,rel2` is working?

And one personal question, how many years of experience do you have in coding? 😄

**Reply by ganigeorgiev** [reply]
Sure. The expand operation happens after we fetch the base record models for the request. 

For each of the fields listed in the `expand` query parameter we are looking for a matching relation field defined in the current collection (this recursive to support nested expansion, eg. relField.subRelField).

Once this field is confirmed that it is a relational, we iterate over each fetched base record to extract the relation ids and then we make a query with those ids (eg. something like `SELECT * from relation where id in (relId1, relId2, ...)`; the query actually is a little more complex because we also attach the related collection view API rule to ensure that only the allowed records are expanded).

And finally we associate manually each record with their expanded relations.

The main code could be found in https://github.com/pocketbase/pocketbase/blob/master/daos/record_expand.go#L43-L156.

**Reply by ganigeorgiev** [reply]
Also, please note that in the next release we'll support "indirect expansion", aka expand of collection records when we don't have direct relation field in the base collection. You could check https://github.com/pocketbase/pocketbase/issues/312

**Comment by tazim404** [comment]
How you are converting 'Go Web App' into an executable?

**Reply by ganigeorgiev** [reply]
PocketBase is like any other Go package and doesn't require anything special to build your Go program. Just run `go build`.

To cross compile from one OS to another you can make use of the `GOOS` and `GOARCH` env variables (preferably with `CGO_ENABLED=0`):

```sh
# compile for mac x86_64
# you can find more info in https://go.dev/doc/install/source#environment
GOARCH=amd64 GOOS=darwin CGO_ENABLED=0 go build main.go
```
